### PR TITLE
Update train.py

### LINKED ISSUE #3: Custom parameters for training
modify code in train.py such that it enables more customizable entry of parameters like epoch and epoch_size

### PR DESCRIPTION
**Pull request by ataha24**
update functions to enable some customizability in training parameters and included validation data functionality

### CODE REVIEW

File: afids_cnn/train.py [hunk 1]
```diff
@@ -6,11 +6,13 @@
 
 import numpy as np
 import pandas as pd
+import tensorflow as tf
 from numpy.typing import NDArray
 from tensorflow import keras
```
Comment (kaitj) [inline afids_cnn/train.py:11]: I don't believe this needs to be imported separately anymore with the change to `tf.keras`

**Review by kaitj — APPROVED** [review]
This looks good to me - there might be 1 import that isn't necessary.

File: afids_cnn/train.py [hunk 2]
```diff
@@ -108,17 +110,32 @@ def gen_model() -> keras.Model:
...
+    epochs: int = 100,
+    steps_per_epoch: int = 50,
+    loss_fn: tf.losses.Loss | str = "mse",
+    optimizer: tf.optimizers.Optimizer | str | None = None,
+    metrics: list[tf.metrics.Metric | str] | None = None,
+    validation_data: Iterable[tuple[NDArray, NDArray]] | None = None,
 ):
+    if not optimizer:
+        optimizer = tf.keras.optimizers.Adam()
+    if metrics is None:
```
Comment (kaitj) [inline afids_cnn/train.py:126]: Just for consistency, can this also be `if not metrics` or does it being a list change how this condition is interpreted

Comment (tkkuehn) [reply afids_cnn/train.py:126]: That would prevent people from legitimately providing an empty list for the metrics. Not sure if that's valid in this framework though.

File: afids_cnn/train.py (hunk 1 above)
Comment (tkkuehn) [reply afids_cnn/train.py:11]: This gets referred to a lot through `tf.losses`, `tf.metrics`, etc. so I think we still need it, but may be wrong

File: afids_cnn/train.py (hunk 2 above)
Comment (kaitj) [reply afids_cnn/train.py:126]: Hmm, that is fair. My thought was that if you aren't providing a metric it'd be equivalent to `None` and should be falling back on whatever the default is in that situation. @ataha24 any insights on this?

Comment (ataha24) [reply afids_cnn/train.py:126]: Here is the documentation reference: https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile. Default is indeed None. Metrics are to be passed as a list, I think. 
Screenshot 2023-09-12 at 11 20 06 AM

File: afids_cnn/train.py (hunk 1 above)
Comment (kaitj) [reply afids_cnn/train.py:11]: Hmm, that feels like that should be handled in the backend by `tf`, rather than requiring an additional import on the end-user.

In any case, I was wrong and this import is still needed as a lot of the typehints use `keras.x` in the output, as well as in `gen_training_array`. Alternatively, can change those to `tf.keras.x` - I think its worth investigating this route just for consistency, but if this import is required after changing those then :see_no_evil:

File: afids_cnn/train.py (hunk 2 above)
Comment (kaitj) [reply afids_cnn/train.py:126]: So by default, `metrics=None`, which would be the same as the `optimizer,` so should be able to handle the condition the same way. Alternatively, the input to `metrics` should always be a list with items in it (according to the typehint), so then this condition would not be met in those cases (and technically make the empty list an invalid input).

I'm not sure what the intended behaviour is suppose to be, because right now if `None` is passed in the workflow, it defaults to RMSE, but if an empty list is passed, I think the model compilation will actually cause tf to error out. If you change this to `if not metrics` then RMSE would be used in the case of both `None` or an empty list being passed in the workflow. _(Hopefully that made sense)_

File: afids_cnn/train.py (hunk 1 above)
Comment (tkkuehn) [reply afids_cnn/train.py:11]: Ok I'll try to get rid of it

File: afids_cnn/train.py (hunk 2 above)
Comment (tkkuehn) [reply afids_cnn/train.py:126]: I'm open to whatever here -- I think the user passing an empty list is kind of an edge case where it's not really clear what the caller is trying to do, so I think just defaulting to RMSE is fine if you want that

Comment (kaitj) [reply afids_cnn/train.py:126]: Yeah fair enough - @ataha24's call

Comment (ataha24) [reply afids_cnn/train.py:126]: ok, I am fine with it defaulting to RMSE to make running workflow easier (i.e., less options to type in) as long as user can override that. Since narritive is landmark regression these metrics (i.e., MSE and RMSE) make sense rather than something like DICE.
